### PR TITLE
[extension/basicauth] Implement HTTP Basic server auth

### PR DESCRIPTION
**Pull request by svrakitin**
**Description:**
As discussed with @jpkrohling in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7165

Implement `configauth.ServerAuthenticator` which matches credentials in Basic Auth format against htpasswd file. It also exposes username as a `username` attribute.

**Documentation:**
Added basic README.md

```yaml
extensions:
  basicauth:
    htpasswd: 
      file: .htpasswd
      inline: |
        ${BASIC_AUTH_USERNAME}:${BASIC_AUTH_PASSWORD}
```

### CODE REVIEW

File: extension/basicauth/README.md
Comment (anuraaga) [inline extension/basicauth/README.md:16]: Should this also accept username / password fields to reference environment variables (and hopefully not embedded directly though it could happen...)? Using extra files like `.htpasswd` can be complicated in environments like k8s

File: extension/basicauth/authdata.go
Comment (anuraaga) [inline extension/basicauth/authdata.go:56]: This is server auth, is it ok to be `client.AuthData`?

Comment (anuraaga) [reply extension/basicauth/authdata.go:56]: No worries if this is unrelated to this PR just wanted to point out this is confusing to me, and in the other function which is also using the `client` package

File: extension/basicauth/authdata.go
Comment (anuraaga) [inline extension/basicauth/authdata.go:31]: I feel this function would be more appropriate within `basicauth.go` - it is the actual authentication business logic, not just a data structure (this file is called `authdata.go`)

File: extension/basicauth/basicauth_test.go
Comment (anuraaga) [inline extension/basicauth/basicauth_test.go:89]: I think it is better to separate out the invalid credentials to its own test case, this is a bit hard to follow otherwise. We don't really need to generate a invalid credential for each of those valid ones

**Review by anuraaga — COMMENTED** [review]
Looks good just some nits

File: extension/basicauth/authdata.go
Comment (svrakitin) [reply extension/basicauth/authdata.go:56]: I took the approach taken in `oidc` extension. I believe the semantics here is "calling client authentication data".

File: extension/basicauth/README.md
Comment (svrakitin) [reply extension/basicauth/README.md:16]: I am not sure how mounting a single file can be complicated, but I can also add an option to inline `.htpasswd` content into the config and supply it through env vars.

Comment (svrakitin) [reply extension/basicauth/README.md:16]: I believe that was the idea of @jpkrohling from https://github.com/open-telemetry/opentelemetry-collector/issues/1424.

**Comment by svrakitin** [comment]
I added an option to inline credentials, which allows sourcing from environment variables.

File: extension/basicauth/README.md
Comment (jpkrohling) [reply extension/basicauth/README.md]: I think we should keep using a file first, as this is what I would expect from an extension like this. If we have requirements from users to embed the contents of the file in the config, we can consider adding that.

Comment (jpkrohling) [reply extension/basicauth/README.md]: I see that you implemented this already. This is probably fine, I'll review it once @anuraaga gives his approval.

**Comment by jpkrohling** [comment]
Could you please go over #6926 and see if you have everything here? I'm willing to be the sponsor for this component.

**Comment by svrakitin** [comment]
@jpkrohling Thanks for the guide. I think I added all the missing pieces.

File: .github/CODEOWNERS
Comment (jpkrohling) [inline .github/CODEOWNERS:60]: The other auth extensions have "authextension" as the suffix. I think it would make sense to follow the pattern for this one as well.

File: extension/basicauth/basicauth.go
Comment (jpkrohling) [inline extension/basicauth/basicauth.go:32]: Ideally, this wouldn't be exported.

**Comment by svrakitin** [comment]
Forgot to add the extension to lifecycle tests.

File: extension/basicauth/basicauth.go
Comment (jpkrohling) [inline extension/basicauth/basicauth.go:39]: Please, use the helper to create the new extension:  `configauth.NewServerAuthenticator`. Like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a395bdfd5f5145672a2f8c73a88f94a15b516c3f/extension/oidcauthextension/extension.go#L70-L75

File: extension/basicauth/README.md
Comment (jpkrohling) [inline extension/basicauth/README.md:3]: Please, state that this authenticator provides the attributes `subject` and `raw` to the client info in the auth data. This is missing from most (all?) current authenticators, but we should add it.

File: extension/basicauth/basicauth.go
Comment (jpkrohling) [inline extension/basicauth/basicauth.go:139]: I think `username` would be more appropriate here instead of `subject`, no? What does httpd use?

File: extension/basicauth/basicauth_test.go
Comment (jpkrohling) [inline extension/basicauth/basicauth_test.go:117]: Can you please add a test that verifies that the inline takes precedence over the file when the same username is present in both?

File: extension/basicauth/basicauth.go
Comment (svrakitin) [reply extension/basicauth/basicauth.go:139]: I thought there is some convention in naming attributes so there is an interoperability between authenticators. I can make it `username` or just make it support both `subject` and `username`.

Comment (jpkrohling) [reply extension/basicauth/basicauth.go:139]: There are different auth standards with slightly different semantics for the names. This is the reason why we have a map and provide the attribute names instead of using a fixed set of attributes and names. So, use whatever is commonly used when configuring htpasswd elsewhere.

File: .github/CODEOWNERS
Comment (svrakitin) [reply .github/CODEOWNERS:60]: I assumed it is unnecessary, because it is already under the `extension` parent, but will rename for the sake of consistency.

File: extension/basicauth/README.md
Comment (svrakitin) [reply extension/basicauth/README.md]: I added the section to the readme file, but probably need help with phrasing.

File: extension/basicauth/basicauth.go
Comment (svrakitin) [reply extension/basicauth/basicauth.go]: Made it just `username`.

File: extension/basicauth/basicauth_test.go
Comment (svrakitin) [reply extension/basicauth/basicauth_test.go]: Done.

File: .github/CODEOWNERS
Comment (svrakitin) [reply .github/CODEOWNERS]: Renamed.

File: extension/basicauthextension/extension.go
Comment (jpkrohling) [inline extension/basicauthextension/extension.go:47]: What happens when no inline data nor file is provided? Will it just reject all requests? I think it should be considered a config error.

**Review by jpkrohling — COMMENTED** [review]
This is looking great!! @pavankrish123, would you like to take a look as well?

File: extension/basicauthextension/extension.go
Comment (svrakitin) [reply extension/basicauthextension/extension.go:47]: It will reject all requests. Can make it fail as you suggested.

Comment (jpkrohling) [reply extension/basicauthextension/extension.go:47]: I think it will be more appropriate.

Comment (svrakitin) [reply extension/basicauthextension/extension.go:58]: Added the check during the creation of the extension.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
@anuraaga, can I get a review from you as well? I don't feel like it's right for me to merge this being the sponsor of this new component.

**Review by anuraaga — APPROVED** [review]

**Comment by svrakitin** [comment]
I wonder if the unit-test job is just flaky or there is some implicit issue introduced with the change. :)

**Comment by svrakitin** [comment]
I just noticed that https://github.com/open-telemetry/opentelemetry-collector/blob/87365b7fde4674d0b50d18a368c775ac05a30576/config/confighttp/clientinfohandler.go#L53 should probably lowercase headers to make it compatible with `configgrpc`. Otherwise this authenticator fails if you pass capitalized header key. I noticed that same is true for `oidc`.

File: extension/basicauthextension/extension.go
Comment (jpkrohling) [inline extension/basicauthextension/extension.go:85]: HTTP headers are case-insensitive, so, the best solution is to iterate over all the headers and use something like strings.EqualFold to check the header name. You can keep those two to short-circuit the checks, like, check first the canonical, then the metadata, then iterate over the headers and do an equal fold.

Comment (svrakitin) [reply extension/basicauthextension/extension.go:85]: Yeah, makes sense. I assumed they will be either canonicalized or lower-cased as coming either from `confighttp` or `configgrpc`.

**Comment by svrakitin** [comment]
I hope that was the last touch. :)

**Review by jpkrohling — APPROVED** [review]

**Comment by svrakitin** [comment]
Thanks @jpkrohling for taking over this and being a sponsor!

**Comment by jpkrohling** [comment]
@svrakitin, could you open a PR with the -releases repository to include this one there?

**Comment by svrakitin** [comment]
@jpkrohling https://github.com/open-telemetry/opentelemetry-collector-releases/pull/62

**Pull request merged**
